platform/speed_formatted.cpp

```diff
@@ -0,0 +1,68 @@
+#include "speed_formatted.hpp"
+
+#include "platform/locale.hpp"
+#include "platform/localization.hpp"
+#include "platform/measurement_utils.hpp"
+
+#include "base/assert.hpp"
+
+namespace platform
+{
+using namespace measurement_utils;
+
+SpeedFormatted::SpeedFormatted(double speedMps) : SpeedFormatted(speedMps, GetMeasurementUnits()) {}
+
+SpeedFormatted::SpeedFormatted(double speedMps, Units units) : m_units(units)
+{
+  switch (units)
+  {
+    case Units::Metric:
+      m_speed = speedMps < 0 ? speedMps : MpsToKmph(speedMps);
+      break;
+    case Units::Imperial:
+      m_speed = speedMps < 0 ? speedMps : MpsToMiph(speedMps);
+      break;
+    default: UNREACHABLE();
+  }
+}
+
+bool SpeedFormatted::IsValid() const { return m_speed >= 0.0; }
+
+double SpeedFormatted::GetSpeed() const { return m_speed; }
+
+Units SpeedFormatted::GetUnits() const { return m_units; }
+
+std::string SpeedFormatted::GetSpeedString() const
+{
+  if (!IsValid())
```

Comment:
What is the situation when this class can be invalid? Can we guarantee that when it is created, it is always valid?

Reply:
Speed limit could be `0.0` which means no limits at all. But if speed limit in OSM data is not set we have `-1.0`. So it makes sense to return `""` is OSM data has no limit defined.

Reply:
Store "invalid" value in valid class doesn't look clean. C++ has std::optional for that, there are also nullptr and empty string that can be used in JNI context to define "missing" speed limit vs "unlimited". AFAIR unlimited roads are only in Germany, and passing there something like 999 should work fine too.